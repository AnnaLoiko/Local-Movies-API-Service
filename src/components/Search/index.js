import React from "react";
import style from "./style.module";

import ButtonPrimary from '@/Components/ButtonPrimary';
import Input from '@/Components/Input';


const Search = () => {
  return (
    <form>
      <div className={style.searchWrap}>
        <Input />
      </div>
      <ButtonPrimary text={"Search"} />
    </form>
  )
}

export default Search;
