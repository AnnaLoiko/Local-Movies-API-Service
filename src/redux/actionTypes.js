const ACTIONS = {
    MOVIE_LOADER: 'MOVIE_LOADER',
    GET_MOVIE_SUCCESS: 'GET_MOVIE_SUCCESS',
    GET_MOVIE_ERROR: 'GET_MOVIE_ERROR',
    ADD_MOVIE_SUCCESS: 'ADD_MOVIE_SUCCESS',
    ADD_MOVIE_ERROR: 'ADD_MOVIE_ERROR',
    EDIT_MOVIE_SUCCESS: 'EDIT_MOVIE_SUCCESS',
    EDIT_MOVIE_ERROR: 'EDIT_MOVIE_ERROR',
    DELETE_MOVIE_SUCCESS: 'DELETE_MOVIE_SUCCESS',
    DELETE_MOVIE_ERROR: 'DELETE_MOVIE_ERROR',
}

export default ACTIONS;

