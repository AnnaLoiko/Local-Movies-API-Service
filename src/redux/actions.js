import axios from 'axios';
import ACTIONS from "./actionTypes";


export const getMovies = (data = {}) => {
  return (dispatch) => {
    dispatch({ type: ACTIONS.MOVIE_LOADER, payload: true });

    axios({
      method: 'GET',
      url: 'http://localhost:4000/movies',
      params: {
        limit: '12',
        filter: (data.filterActiveKey === 'All') ? '' : data.filterActiveKey,
        sortBy: data.sortActiveKey,
        sortOrder: data.sortOrder,
      },
    })
      .then(response => {
        dispatch({
          type: ACTIONS.GET_MOVIE_SUCCESS,
          payloadResponse: response.data.data,
          payloadParams: {
            filterActiveKey: data.filterActiveKey || 'All',
            sortActiveKey: data.sortActiveKey || 'release_date',
            sortOrder: data.sortOrder,
          }
        });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.GET_MOVIE_ERROR, payload: error });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
  }
};

export const addMovie = (newMovie) => {
  return (dispatch) => {
    axios({
      method: 'POST',
      url: 'http://localhost:4000/movies',
      data: newMovie,
    })
      .then(response => {
        dispatch({
          type: ACTIONS.ADD_MOVIE_SUCCESS,
          payload: {...newMovie, id: response.data.id },
        });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.ADD_MOVIE_ERROR, payload: error });
      })
  }
};

export const editMovie = (editedMovie) => {
  return (dispatch) => {
    axios({
      method: 'PUT',
      url: 'http://localhost:4000/movies',
      data: editedMovie,
    })
      .then(response => {
        dispatch({
          type: ACTIONS.EDIT_MOVIE_SUCCESS,
          payload: response.data,
        });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.EDIT_MOVIE_ERROR, payload: error });
      })
  }
};

export const deleteMovie = (movieId) => {
  return (dispatch) => {
    axios({
      method: 'DELETE',
      url: `http://localhost:4000/movies/${movieId}`,
      params: {
        id: movieId,
      },
    })
      .then(response => {
        console.log('response', response);
        dispatch({
          type: ACTIONS.DELETE_MOVIE_SUCCESS,
          payload: movieId,
        });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.DELETE_MOVIE_ERROR, payload: error });
      })
  }
};

export const getMovieById = (movieId) => {
  return (dispatch) => {
    axios({
      method: 'GET',
      url: `http://localhost:4000/movies/${movieId}`,
      params: {
        id: movieId,
      },
    })
      .then((response) => {
        dispatch({
          type: ACTIONS.GET_MOVIE_BY_ID_SUCCESS,
          payload: response.data,
        });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.GET_MOVIE_BY_ID_ERROR, payload: error});
      })
  }
};

export const postMessage = () => ({type: ACTIONS.PUT_MESSAGE_ADD_MOVIE_SUCCESS, payload: false });
