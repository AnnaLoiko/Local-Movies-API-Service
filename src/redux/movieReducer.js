import ACTIONS from "./actionTypes";

const initialState = {
  movies: [],
  currentMovie: {},
  params: {
    filterActiveKey: "All",
    sortActiveKey: "release_date",
    sortOrder: '',
  },
  filterKeys: ["All", 'Documentary', 'Comedy', 'Horror', 'Crime'],
  sortByKeys: [
    { key: "release_date", title: "Release date", sortOrder: "desc", isSelected: false },
    { key: "release_date", title: "Release date", sortOrder: "asc", isSelected: false },
    { key: "vote_average", title: "Rating", sortOrder: "desc", isSelected: false },
  ],
  genresList: ['Documentary', 'Comedy', 'Horror', 'Crime', 'Animation', 'Adventure', 'Family', 'Drama', 'Romance', 'Fantasy', 'Adventure', 'Science Fiction'],
  loader: false,
  messageAddMovieSucc: false,
  messageEditMovieSucc: false,
  errorGetMovie: false,
  
  // not use yet
  errorAddMovie: false,
  errorEditMovie: false,
  errorDeleteMovie: false,
  errorGetMovieById: false,
}

const movieReducer = (state = initialState, action) => {
  switch (action.type) {

    case ACTIONS.MOVIE_LOADER:
      return {
        ...state,
        loader: action.payload,
      };

    case ACTIONS.GET_MOVIE_SUCCESS:
      return {
        ...state,
        movies: action.payloadResponse,
        params: {
          ...state.params,
          ...action.payloadParams,
        },
        errorGetMovie: false,
      };

    case ACTIONS.GET_MOVIE_ERROR:
      return {
        ...state,
        errorGetMovie: true,
      };

    case ACTIONS.DELETE_MOVIE_SUCCESS:
      return {
        ...state,
        movies: state.movies.filter(item => item.id !== action.payload),
        errorDeleteMovie: false,
      };

    case ACTIONS.DELETE_MOVIE_ERROR:
      return {
        ...state,
        errorDeleteMovie: true,
      };

    case ACTIONS.ADD_MOVIE_SUCCESS:
      return {
        ...state,
        movies: [
          { ...action.payload},
          ...state.movies
        ],
        errorAddMovie: false,
        messageAddMovieSucc: true,
      };

    case ACTIONS.PUT_MESSAGE_ADD_MOVIE_SUCCESS:
      return {
        ...state,
        messageAddMovieSucc: action.payload,
      };

    case ACTIONS.ADD_MOVIE_ERROR:
      return {
        ...state,
        errorAddMovie: true,
      };

    case ACTIONS.EDIT_MOVIE_SUCCESS:
      return {
        ...state,
        movies: state.movies.map(
          item =>
            item.id === action.payload.id
              ? { ...item, ...action.payload }
              : item
        ),
        errorEditMovie: false,
        messageEditMovieSucc: true,
      };

    case ACTIONS.EDIT_MOVIE_ERROR:
      return {
        ...state,
        errorEditMovie: true,
      };

    case ACTIONS.GET_MOVIE_BY_ID_SUCCESS:
      return {
        ...state,
        currentMovie: { ...action.payload },
        errorGetMovieById: false,
      };

    case ACTIONS.GET_MOVIE_BY_ID_ERROR:
      return {
        ...state,
        errorGetMovieById: true,
      };

    default:
      return state;

  }
}

export default movieReducer;