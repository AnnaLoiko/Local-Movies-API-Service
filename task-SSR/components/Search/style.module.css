.searchWrap {
    width: 70%;
    display: inline-block;
    margin-right: 20px;
}