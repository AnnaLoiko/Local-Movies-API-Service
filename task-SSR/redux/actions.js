import axios from 'axios';

import * as api from '../api';
import ACTIONS from "./actionTypes";

export const getMovies = (data = {}) => {
  return (dispatch) => {
    dispatch({ type: ACTIONS.MOVIE_LOADER, payload: true });

    
    return api.getMovies(data)
      .then(response => {
        console.log('!!!!!!!!!!______response:', data)
        console.log('!!!!!!!!!!______response:', response)
        dispatch({
          type: ACTIONS.GET_MOVIE_SUCCESS,
          payloadResponse: response.data.data,
        });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.GET_MOVIE_ERROR, payload: error });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
  }
};

export const addMovie = (newMovie) => {
  return (dispatch) => {
    return axios({
      method: 'POST',
      url: 'http://localhost:4000/movies',
      data: newMovie,
    })
      .then(response => {
        dispatch({
          type: ACTIONS.ADD_MOVIE_SUCCESS,
          payload: {...newMovie, id: response.data.id },
        });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.ADD_MOVIE_ERROR, payload: error });
      })
  }
};

export const editMovie = (editedMovie) => {
  return (dispatch) => {
    return axios({
      method: 'PUT',
      url: 'http://localhost:4000/movies',
      data: editedMovie,
    })
      .then(response => {
        dispatch({
          type: ACTIONS.EDIT_MOVIE_SUCCESS,
          payload: response.data,
        });
        dispatch({ type: ACTIONS.MOVIE_LOADER, payload: false });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.EDIT_MOVIE_ERROR, payload: error });
      })
  }
};

export const deleteMovie = (movieId) => {
  return (dispatch) => {
    return axios({
      method: 'DELETE',
      url: `http://localhost:4000/movies/${movieId}`,
      params: {
        id: movieId,
      },
    })
      .then(() => {
        dispatch({
          type: ACTIONS.DELETE_MOVIE_SUCCESS,
          payload: movieId,
        });
      })
      .catch((error) => {
        dispatch({ type: ACTIONS.DELETE_MOVIE_ERROR, payload: error });
      })
  }
};

export const postMessage = () => ({type: ACTIONS.PUT_MESSAGE_ADD_MOVIE_SUCCESS, payload: false });

export const resetSearchData = () => ({type: ACTIONS.RESET_SEARCH_DATA});

export const updateParams = (params) => ({
  type: ACTIONS.UPDATE_PARAMS,
  params,
});

export const resetParams = () => ({
  type: ACTIONS.RESET_PARAMS,
});
