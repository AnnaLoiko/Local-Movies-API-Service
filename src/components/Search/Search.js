import React, { useState, useEffect  } from "react";
import { connect } from 'react-redux';
import { useHistory, useLocation } from "react-router-dom";

import style from "./style.module";

import Input from '@/Components/Input/Input';
import Button from '@/Components/Button/Button';

import { getMovies } from "@/redux/actions";
import useUpdateEffect from '@/hooks/useUpdateEffect';

const Search = (props) =>  {
  const { params, getMovies } = props;

  const [value, setValue] = useState('');
  const history = useHistory();

  const searchChange = (e) => {
    e.preventDefault(),
    getMovies({...params, search: value}),
    history.push(`/search/Search%20${value}`)
  }

  useUpdateEffect(() => {
    params && !params.search && setValue('');
  }, [params && params.search])

  return (
    <form>
      <div className={style.searchWrap}>
        <Input
          type="search"
          placeholder="What do you want to watch?"
          value={value}
          onChange={e => {setValue(e.target.value)}}
          role='textbox'
        />
      </div>
      
      <Button 
        text="Search"
        type="text"
        className="btnPrimary"
        addClassName="btnLarge"
        handleClick={(e) => searchChange(e)}
        role="searchButton"
      />
    </form>
  )
}

const mapStateToProps = state => {
  return {
    params: state.movies.params,
  }
}

const mapDispatchToProps = {
  getMovies
}

export default connect(mapStateToProps, mapDispatchToProps)(Search);