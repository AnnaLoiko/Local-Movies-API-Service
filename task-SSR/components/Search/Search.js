import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from 'react-redux';
import { useRouter } from "next/router";

import * as actions from "@/redux/actions";

import Input from '@/Components/Input/Input';
import Button from '@/Components/Button/Button';

import style from "./style.module";

const Search = () =>  {
  const router = useRouter();
  const dispatch = useDispatch();

  const params = useSelector((state) => state.movies.params);

  const [value, setValue] = useState();

  useEffect(() => {
    setValue(params.search);
  }, [params.search]);

  const searchChange = (e) => {
    e.preventDefault();

    dispatch(actions.updateParams({ search: value }));

    router.replace(`/?search=${value}`);
  };

  return (
    <form>
      <div className={style.searchWrap}>
        <Input
          type="search"
          placeholder="What do you want to watch?"
          value={value}
          onChange={e => {setValue(e.target.value)}}
          role='textbox'
        />
      </div>

      <Button
        text="Search"
        type="text"
        className="btnPrimary"
        addClassName="btnLarge"
        handleClick={(e) => searchChange(e)}
        role="searchButton"
      />
    </form>
  )
}

export default Search;
