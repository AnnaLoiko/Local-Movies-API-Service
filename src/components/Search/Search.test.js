import React from "react";
import { Provider } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import * as actions from '@/redux/actions';

import renderer from 'react-test-renderer';

import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);
import userEvent from '@testing-library/user-event';

import Search from "./Search";
// import ACTIONS from "../../redux/actionTypes";

describe('Search  /  Connected React-Redux Component Search: ', () => {

    let store;
    let component;
    let onChangeMock;
   
    beforeEach(() => {
      store = mockStore({
        movies: [],
        params: {},
      });


      onChangeMock = jest.fn();
   
      component = render(
        <Provider store={store}>
          <Search value="" onChange={onChangeMock} />
        </Provider>
      );
    });

    it('Should render snapshot Search component', () => {
        const { asFragment } = component;
        expect( asFragment(<Search />) ).toMatchSnapshot();
    });

    it('Find elements input Search by placeholder', () => {
      expect(screen.getByPlaceholderText('What do you want to watch?')).toBeInTheDocument();
    });

    it('Search - userEvent', () => {
        userEvent.type(screen.getByRole('textbox'), "Javascript");
        expect(screen.getByDisplayValue("Javascript")).toBeInTheDocument();
    });








    // it('Calls the onChange callback handler', () => {
    //     userEvent.type(screen.getByRole('textbox'), "Javascript");
    //     expect(onChangeMock).toHaveBeenCalledTimes(1);
    // });

   
    // it('should dispatch an action on button click', () => {

    //     const testRenderer = renderer.create(
    //         <Provider store={store}>
    //         <Search value="" onChange={onChange} />
    //       </Provider>
    //       );
    //       const testInstance = testRenderer.root;

    //     renderer.act(() => {
    //         testInstance.findByType('button').props.onClick();
    //     });
     
    //     expect(store.dispatch).toHaveBeenCalledTimes(1);
    //     expect(store.dispatch).toHaveBeenCalledWith(
    //         actions.getMovies({ payload: "Javascript" })
    //     );
    // });
})