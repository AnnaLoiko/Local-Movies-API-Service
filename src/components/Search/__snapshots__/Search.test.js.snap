// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search  /  Connected React-Redux Component Search:  Should render snapshot Search component 1`] = `
<DocumentFragment>
  <form>
    <div
      class="searchWrap"
    >
      <input
        class="input undefined"
        placeholder="What do you want to watch?"
        role="textbox"
        type="search"
        value=""
      />
    </div>
    <button
      class="btnPrimary btnLarge "
      type="text"
    >
      <span>
        Search
      </span>
    </button>
  </form>
</DocumentFragment>
`;
